File: lib/domain/entities/medicine_entity.dart
import 'package:equatable/equatable.dart';

class MedicineEntity extends Equatable {
  final String id;
  final String userId;
  final String name;
  final String dosage;
  final String unit;
  final String frequency;
  final List<String> timeSlots;
  final DateTime startDate;
  final DateTime? endDate;
  final bool isActive;
  final String? notes;
  final DateTime createdAt;
  final DateTime updatedAt;

  const MedicineEntity({
    required this.id,
    required this.userId,
    required this.name,
    required this.dosage,
    required this.unit,
    required this.frequency,
    required this.timeSlots,
    required this.startDate,
    this.endDate,
    this.isActive = true,
    this.notes,
    required this.createdAt,
    required this.updatedAt,
  });

  @override
  List<Object?> get props => [
        id,
        userId,
        name,
        dosage,
        unit,
        frequency,
        timeSlots,
        startDate,
        endDate,
        isActive,
        notes,
        createdAt,
        updatedAt,
      ];

  MedicineEntity copyWith({
    String? id,
    String? userId,
    String? name,
    String? dosage,
    String? unit,
    String? frequency,
    List<String>? timeSlots,
    DateTime? startDate,
    DateTime? endDate,
    bool? isActive,
    String? notes,
    DateTime? createdAt,
    DateTime? updatedAt,
  }) {
    return MedicineEntity(
      id: id ?? this.id,
      userId: userId ?? this.userId,
      name: name ?? this.name,
      dosage: dosage ?? this.dosage,
      unit: unit ?? this.unit,
      frequency: frequency ?? this.frequency,
      timeSlots: timeSlots ?? this.timeSlots,
      startDate: startDate ?? this.startDate,
      endDate: endDate ?? this.endDate,
      isActive: isActive ?? this.isActive,
      notes: notes ?? this.notes,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
    );
  }
}

File: backend/src/medicines/entities/medicine.entity.ts
import { ApiProperty, ApiPropertyOptional } from '@nestjs/swagger';

export enum MedicineStatus {
  ACTIVE = 'active',
  COMPLETED = 'completed',
  PAUSED = 'paused',
  STOPPED = 'stopped',
}

export enum MedicineUnit {
  TABLET = 'tablet',
  CAPSULE = 'capsule',
  ML = 'ml',
  MG = 'mg',
  DROPS = 'drops',
  INJECTION = 'injection',
}

export class Medicine {
  @ApiProperty()
  id: string;

  @ApiProperty()
  userId: string;

  @ApiPropertyOptional()
  familyMemberId?: string;

  @ApiProperty({ example: 'Aspirin' })
  medicineName: string;

  @ApiProperty({ example: 500 })
  dosage: number;

  @ApiProperty({ enum: MedicineUnit })
  unit: MedicineUnit;

  @ApiProperty({ example: 'Twice daily' })
  frequency: string;

  @ApiProperty()
  startDate: Date;

  @ApiPropertyOptional()
  endDate?: Date;

  @ApiPropertyOptional({ type: [String] })
  times?: string[];

  @ApiPropertyOptional()
  instructions?: string;

  @ApiProperty({ enum: MedicineStatus })
  status: MedicineStatus;

  @ApiProperty()
  createdAt: Date;

  @ApiProperty()
  updatedAt: Date;
}
